components/hab/src/main.rs

```diff
@@ -1490,6 +1501,51 @@ where
     return Ok(());
 }
 
+/// Check if we have a launcher/supervisor running out of this habitat root.
+/// If the launcher PID file exists then the supervisor is up and running
+fn launcher_is_running(fs_root_path: &Path) -> bool {
+    let launcher_root = launcher_root_path(Some(fs_root_path));
+    let pid_file_path = launcher_root.join("PID");
+
+    pid_file_path.is_file()
+}
+
+fn supervisor_services() -> Result<Option<Vec<PackageIdent>>> {
+    if !launcher_is_running(&*FS_ROOT) {
+        return Ok(None);
+    }
+
+    let cfg = config::load()?;
+    let secret_key = ctl_secret_key(&cfg)?;
+    let sup_addr = protocol::ctl::default_addr();
+    let msg = protocol::ctl::SvcStatus::default();
+
+    let mut out: Vec<PackageIdent> = vec![];
+    SrvClient::connect(&sup_addr, &secret_key)
+        .and_then(|conn| {
+            conn.call(msg).for_each(|reply| {
+                match reply.message_id() {
+                    "ServiceStatus" => {
+                        let m = reply.parse::<protocol::types::ServiceStatus>().unwrap();
+                        out.push(m.ident.into());
+                    }
+                    "NetOk" => {
+                        // All Ok, no services loaded
+                    }
+                    "NetErr" => {
+                        let err = reply.parse::<protocol::net::NetErr>().unwrap();
+                        return Err(SrvClientError::from(err));
+                    }
+                    _ => {
+                        warn!("Unexpected status message, {:?}", reply);
+                    }
+                };
+                Ok(())
```

Comment:
I think it's a little easier to follow to have each `match` arm return the appropriate `Result` value than fall through here.